### PR TITLE
[S3 API] Support for S3 AbortMultipartUpload API

### PR DESCRIPTION
**Pull request by alyssaxu333**
## Summary
S3 Abort Multipart Upload API support
https://docs.aws.amazon.com/AmazonS3/latest/API/API_AbortMultipartUpload.html

## Testing Done
Unit test, gradlew build

### CODE REVIEW

**Comment by justinlin-linkedin** [comment]
Could you also change your title to be more descriptive?

File: ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java [hunk 1]
```diff
@@ -229,12 +231,13 @@ public void start() throws InstantiationException {
     postAccountsHandler = new PostAccountsHandler(securityService, accountService, frontendConfig, frontendMetrics);
     postDatasetsHandler = new PostDatasetsHandler(securityService, accountService, frontendConfig, frontendMetrics,
         accountAndContainerInjector);
-    s3DeleteHandler = new S3DeleteHandler(deleteBlobHandler, frontendMetrics);
     s3HeadHandler = new S3HeadHandler(headBlobHandler, securityService, frontendMetrics, accountService);
     s3HeadHandler = new S3HeadHandler(headBlobHandler, securityService, frontendMetrics, accountService);
```
[the remark this answers is not in the document]
Comment (alyssaxu333) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java:235]: I didn't notice, just deleted.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java [hunk 2]
```diff
@@ -229,12 +231,13 @@ public void start() throws InstantiationException {
...
     s3MultipartUploadHandler =
         new S3MultipartUploadHandler(securityService, frontendMetrics, accountAndContainerInjector, frontendConfig,
             namedBlobDb, idConverter, router, quotaManager);
+    s3MultipartAbortHandler = new S3MultipartAbortUploadHandler(securityService, frontendMetrics, accountAndContainerInjector);
+    s3DeleteHandler = new S3DeleteHandler(deleteBlobHandler, s3MultipartAbortHandler, frontendMetrics);
```
[the remark this answers is not in the document]
Comment (alyssaxu333) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java:240]: Yea I originally planned to do that. But s3MultipartUploadHandler doHandle() pass in a Callback<ReadableStreamChannel> , while s3DeleteHandler doHandle() passes in a Callback<Void>. If we wanna pass in s3MultipartUploadHandler, probably need to do more changes to s3DeleteHandler & deleteBlobHandler

[the remark this answers is not in the document]
Comment (Arun-LinkedIn) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java:240]: I think if we use generic type parameter `R` in `S3MultipartUploadHandler` class, it allows us to pass S3MultipartUploadHandler class in both `S3DeleteHandler` and existing classes. Something like below?

```
public class S3MultipartUploadHandler<R> extends S3BaseHandler<R>
.....
    
    abortUploadHandler =
        new S3MultipartAbortUploadHandler(securityService, frontendMetrics, accountAndContainerInjector);
.....


  @Override
  @SuppressWarnings("unchecked")
  protected void doHandle(RestRequest restRequest, RestResponseChannel restResponseChannel, Callback<R> callback)
      throws RestServiceException {
    if (isMultipartCreateUploadRequest(restRequest)) {
      createMultipartUploadHandler.handle(restRequest, restResponseChannel, (Callback<ReadableStreamChannel>) callback);
    } else if(isMultipartAbortUploadRequest(restRequest)){
      abortUploadHandler.handle(restRequest, restResponseChannel, (Callback<Void>) callback);
    }
      
```

[the remark this answers is not in the document]
Comment (alyssaxu333) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java:240]: Thanks! I'll revise

File: ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3DeleteHandler.java [hunk 3]
```diff
@@ -57,6 +62,9 @@ public S3DeleteHandler(DeleteBlobHandler deleteBlobHandler, FrontendMetrics metr
    */
   protected void doHandle(RestRequest restRequest, RestResponseChannel restResponseChannel, Callback<Void> callback)
       throws RestServiceException {
+    if(S3MultipartUploadHandler.isMultipartAbortUploadRequest(restRequest)) {
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3DeleteHandler.java:65]: Again, this is not resolved.

[the remark this answers is not in the document]
Comment (alyssaxu333) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3DeleteHandler.java:65]: Done

**Review by justinlin-linkedin — APPROVED** [review]

**Review by Arun-LinkedIn — APPROVED** [review]

**Pull request merged**